File: scripts/analyze_positions.js
#!/usr/bin/env node

/**
 * Script to analyze open positions, compare DB vs Exchange, and find losing positions
 */

import mysql from 'mysql2/promise';
import { config } from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { readFileSync } from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load .env
config({ path: join(__dirname, '..', '.env') });

async function getDbConnection() {
  return mysql.createPool({
    host: process.env.DB_HOST || 'localhost',
    user: process.env.DB_USER || 'root',
    password: process.env.DB_PASSWORD || '',
    database: process.env.DB_NAME || 'bot_oc',
    waitForConnections: true,
    connectionLimit: 10
  });
}

async function analyzePositions() {
  const pool = await getDbConnection();
  
  try {
    console.log('='.repeat(80));
    console.log('📊 POSITION ANALYSIS REPORT');
    console.log('='.repeat(80));
    console.log(`Generated at: ${new Date().toISOString()}\n`);
    
    // 1. Get open positions from database
    console.log('1️⃣  Fetching open positions from database...');
    // Get all columns from positions, then join for additional info
    const [dbPositions] = await pool.execute(`
      SELECT 
        p.*,
        b.exchange,
        b.binance_testnet,
        s.extend,
        s.take_profit,
        s.stoploss
      FROM positions p
      LEFT JOIN bots b ON p.bot_id = b.id
      LEFT JOIN strategies s ON p.strategy_id = s.id
      WHERE p.status = 'open'
      ORDER BY p.opened_at DESC
    `);
    
    console.log(`   Found ${dbPositions.length} open positions in database\n`);
    
    if (dbPositions.length === 0) {
      console.log('✅ No open positions found. System is clean!\n');
      return;
    }
    
    // 2. Analyze by exchange
    console.log('2️⃣  Analyzing by exchange...');
    const byExchange = {};
    dbPositions.forEach(pos => {
      const exchange = pos.exchange || 'unknown';
      const isTestnet = pos.binance_testnet;
      const key = `${exchange}${isTestnet ? '_testnet' : ''}`;
      if (!byExchange[key]) byExchange[key] = [];
      byExchange[key].push(pos);
    });
    
    Object.entries(byExchange).forEach(([exchange, positions]) => {
      console.log(`   ${exchange}: ${positions.length} positions`);
    });
    console.log();
    
    // Calculate PnL for each position (if not in DB)
    // Note: PnL is calculated in PositionService.updatePosition() but may not be stored in DB
    // We'll fetch current prices and calculate PnL for analysis
    console.log('   Calculating current PnL for positions...');
    // For now, we'll use pnl from DB if available, otherwise mark as unknown
    dbPositions.forEach(pos => {
      // pnl_percent calculation: ((current_price - entry_price) / entry_price) * 100
      // But we need current price, so we'll use pnl from DB if available
      if (pos.pnl === undefined || pos.pnl === null) {
        pos.pnl = null;
        pos.pnl_percent = null;
      } else {
        // Calculate pnl_percent if not available
        if (pos.pnl_percent === undefined || pos.pnl_percent === null) {
          if (pos.entry_price && pos.amount && pos.entry_price > 0) {
            // pnl_percent = (pnl / amount) * 100
            pos.pnl_percent = (pos.pnl / pos.amount) * 100;
          } else {
            pos.pnl_percent = 0;
          }
        }
      }
    });
    
    // 3. Analyze losing positions
    console.log('3️⃣  Analyzing losing positions...');
    const losingPositions = dbPositions.filter(p => p.pnl !== null && p.pnl < 0);
    console.log(`   Total losing positions: ${losingPositions.length}`);
    
    if (losingPositions.length > 0) {
      const totalLoss = losingPositions.reduce((sum, p) => sum + (Number(p.pnl) || 0), 0);
      const avgLoss = totalLoss / losingPositions.length;
      const maxLoss = Math.min(...losingPositions.map(p => Number(p.pnl) || 0));
      
      console.log(`   Total loss: ${totalLoss.toFixed(2)} USDT`);
      console.log(`   Average loss: ${avgLoss.toFixed(2)} USDT`);
      console.log(`   Max loss: ${maxLoss.toFixed(2)} USDT`);
      
      console.log('\n   Top 10 losing positions:');
      const sortedLosing = losingPositions
        .sort((a, b) => (a.pnl || 0) - (b.pnl || 0))
        .slice(0, 10);
      
      sortedLosing.forEach((pos, i) => {
        const hoursOpen = pos.opened_at 
          ? ((Date.now() - new Date(pos.opened_at).getTime()) / (1000 * 60 * 60)).toFixed(1)
          : 'N/A';
        const pnl = pos.pnl !== null && pos.pnl !== undefined ? Number(pos.pnl) : null;
        const pnlPercent = pos.pnl_percent !== null && pos.pnl_percent !== undefined ? Number(pos.pnl_percent) : null;
        console.log(`   ${i + 1}. ${pos.symbol} ${(pos.side || '').toUpperCase()} | ` +
                   `Entry: ${Number(pos.entry_price || 0).toFixed(8)} | ` +
                   `PnL: ${pnl !== null ? pnl.toFixed(2) : 'N/A'} USDT ${pnlPercent !== null ? '(' + pnlPercent.toFixed(2) + '%)' : ''} | ` +
                   `SL: ${pos.stop_loss_price ? Number(pos.stop_loss_price).toFixed(8) : 'N/A'} | ` +
                   `Open: ${hoursOpen}h`);
      });
    }
    console.log();
    
    // 4. Analyze winning positions
    console.log('4️⃣  Analyzing winning positions...');
    const winningPositions = dbPositions.filter(p => p.pnl !== null && p.pnl > 0);
    console.log(`   Total winning positions: ${winningPositions.length}`);
    
    if (winningPositions.length > 0) {
      const totalProfit = winningPositions.reduce((sum, p) => sum + (Number(p.pnl) || 0), 0);
      const avgProfit = totalProfit / winningPositions.length;
      const maxProfit = Math.max(...winningPositions.map(p => Number(p.pnl) || 0));
      
      console.log(`   Total profit: ${totalProfit.toFixed(2)} USDT`);
      console.log(`   Average profit: ${avgProfit.toFixed(2)} USDT`);
      console.log(`   Max profit: ${maxProfit.toFixed(2)} USDT`);
    }
    console.log();
    
    // 5. Analyze by strategy type
    console.log('5️⃣  Analyzing by strategy type...');
    const byStrategyType = {};
    
    dbPositions.forEach(pos => {
      // Check if strategy has is_reverse_strategy field (may be in bots table)
      const isReverse = pos.is_reverse_strategy || false;
      const strategyType = isReverse ? 'COUNTER_TREND' : 'FOLLOWING_TREND';
      if (!byStrategyType[strategyType]) {
        byStrategyType[strategyType] = { total: 0, winning: 0, losing: 0, totalPnl: 0 };
      }
      byStrategyType[strategyType].total++;
      const pnl = pos.pnl !== null && pos.pnl !== undefined ? Number(pos.pnl) : 0;
      byStrategyType[strategyType].totalPnl += pnl;
      if (pnl > 0) byStrategyType[strategyType].winning++;
      else if (pnl < 0) byStrategyType[strategyType].losing++;
    });
    
    Object.entries(byStrategyType).forEach(([type, stats]) => {
      const winRate = stats.total > 0 ? (stats.winning / stats.total * 100) : 0;
      console.log(`   ${type}:`);
      console.log(`      Total: ${stats.total} | Winning: ${stats.winning} | Losing: ${stats.losing}`);
      console.log(`      Win Rate: ${winRate.toFixed(2)}% | Total PnL: ${stats.totalPnl.toFixed(2)} USDT`);
    });
    console.log();
    
    // 6. Analyze positions without SL
    console.log('6️⃣  Analyzing positions without Stop Loss...');
    const noSlPositions = dbPositions.filter(p => !p.stop_loss_price || Number(p.stop_loss_price) === 0);
    console.log(`   Positions without SL: ${noSlPositions.length}`);
    if (noSlPositions.length > 0) {
      console.log('   ⚠️  WARNING: These positions are at risk!');
      noSlPositions.slice(0, 10).forEach(pos => {
        console.log(`      - ${pos.symbol} ${pos.side.toUpperCase()} | ` +
                   `Entry: ${Number(pos.entry_price).toFixed(8)} | ` +
                   `PnL: ${(pos.pnl || 0).toFixed(2)} USDT | ` +
                   `Strategy SL: ${pos.stoploss || 'N/A'}`);
      });
    }
    console.log();
    
    // 7. Analyze positions without TP
    console.log('7️⃣  Analyzing positions without Take Profit...');
    const noTpPositions = dbPositions.filter(p => !p.take_profit_price || Number(p.take_profit_price) === 0);
    console.log(`   Positions without TP: ${noTpPositions.length}`);
    if (noTpPositions.length > 0) {
      console.log('   ⚠️  WARNING: These positions may not exit at profit target!');
      noTpPositions.slice(0, 10).forEach(pos => {
        console.log(`      - ${pos.symbol} ${pos.side.toUpperCase()} | ` +
                   `Entry: ${Number(pos.entry_price).toFixed(8)} | ` +
                   `PnL: ${(pos.pnl || 0).toFixed(2)} USDT`);
      });
    }
    console.log();
    
    // 8. Analyze by time opened
    console.log('8️⃣  Analyzing by time opened...');
    const now = Date.now();
    const recentPositions = [];
    const oldPositions = [];
    
    dbPositions.forEach(pos => {
      if (pos.opened_at) {
        const openedAt = new Date(pos.opened_at).getTime();
        const hoursOpen = (now - openedAt) / (1000 * 60 * 60);
        if (hoursOpen < 24) {
          recentPositions.push({ pos, hours: hoursOpen });
        } else {
          oldPositions.push({ pos, hours: hoursOpen });
        }
      }
    });
    
    console.log(`   Positions opened < 24h: ${recentPositions.length}`);
    console.log(`   Positions opened >= 24h: ${oldPositions.length}`);
    
    if (oldPositions.length > 0) {
      console.log('   ⚠️  WARNING: Old positions that may need review:');
      oldPositions
        .sort((a, b) => b.hours - a.hours)
        .slice(0, 5)
        .forEach(({ pos, hours }) => {
          console.log(`      - ${pos.symbol} ${pos.side.toUpperCase()} | ` +
                     `Open for ${hours.toFixed(1)}h | ` +
                     `PnL: ${(pos.pnl || 0).toFixed(2)} USDT`);
        });
    }
    console.log();
    
    // 9. Analyze positions with large losses
    console.log('9️⃣  Analyzing positions with large losses (>5% or >50 USDT)...');
    const largeLossPositions = losingPositions.filter(p => {
      const pnlPercent = Math.abs(p.pnl_percent !== null ? p.pnl_percent : 0);
      const pnlAbs = Math.abs(p.pnl !== null ? p.pnl : 0);
      return pnlPercent > 5 || pnlAbs > 50;
    });
    console.log(`   Positions with large losses: ${largeLossPositions.length}`);
    if (largeLossPositions.length > 0) {
      console.log('   ⚠️  CRITICAL: These positions need immediate attention!');
      largeLossPositions.slice(0, 10).forEach(pos => {
        console.log(`      - ${pos.symbol} ${pos.side.toUpperCase()} | ` +
                   `Entry: ${Number(pos.entry_price).toFixed(8)} | ` +
                   `PnL: ${pos.pnl !== null && pos.pnl !== undefined ? Number(pos.pnl).toFixed(2) : 'N/A'} USDT ${pos.pnl_percent !== null && pos.pnl_percent !== undefined ? '(' + Number(pos.pnl_percent).toFixed(2) + '%)' : ''} | ` +
                   `SL: ${pos.stop_loss_price ? Number(pos.stop_loss_price).toFixed(8) : 'MISSING!'}`);
      });
    }
    console.log();
    
    // 10. Summary
    console.log('='.repeat(80));
    console.log('📊 SUMMARY');
    console.log('='.repeat(80));
    const totalPnl = dbPositions.reduce((sum, p) => sum + (p.pnl !== null && p.pnl !== undefined ? Number(p.pnl) : 0), 0);
    const winRate = dbPositions.length > 0 
      ? (winningPositions.length / dbPositions.length * 100) 
      : 0;
    
    console.log(`Total Open Positions: ${dbPositions.length}`);
    console.log(`Winning: ${winningPositions.length} | Losing: ${losingPositions.length}`);
    console.log(`Total PnL: ${totalPnl.toFixed(2)} USDT`);
    console.log(`Win Rate: ${winRate.toFixed(2)}%`);
    console.log(`Positions without SL: ${noSlPositions.length}`);
    console.log(`Positions without TP: ${noTpPositions.length}`);
    console.log(`Positions with large losses: ${largeLossPositions.length}`);
    console.log();
    
    // 11. Recommendations
    console.log('='.repeat(80));
    console.log('💡 RECOMMENDATIONS');
    console.log('='.repeat(80));
    
    if (noSlPositions.length > 0) {
      console.log('⚠️  CRITICAL: Some positions don\'t have Stop Loss!');
      console.log('   → Check PositionMonitor logs for SL placement errors');
      console.log('   → Verify strategy.stoploss is configured');
      console.log('   → Check ExitOrderManager logs');
    }
    
    if (noTpPositions.length > 0) {
      console.log('⚠️  WARNING: Some positions don\'t have Take Profit!');
      console.log('   → Check PositionMonitor logs for TP placement errors');
      console.log('   → Verify strategy.take_profit is configured');
    }
    
    if (losingPositions.length > winningPositions.length) {
      console.log('⚠️  WARNING: More losing positions than winning!');
      console.log('   → Review entry conditions and trend filters');
      console.log('   → Check if SL is being hit too early');
      console.log('   → Consider adjusting strategy parameters');
    }
    
    if (largeLossPositions.length > 0) {
      console.log('🚨 CRITICAL: Positions with large losses detected!');
      console.log('   → Review these positions immediately');
      console.log('   → Check if SL orders are working correctly');
      console.log('   → Verify exchange position status');
    }
    
    if (oldPositions.length > 0) {
      console.log('⚠️  INFO: Some positions are open for > 24h');
      console.log('   → Review if these should still be open');
      console.log('   → Check trailing TP logic');
      console.log('   → Consider manual review');
    }
    
    console.log();
    console.log('='.repeat(80));
    console.log('✅ Analysis complete!');
    console.log('='.repeat(80));
    
  } finally {
    await pool.end();
  }
}

// Run analysis
analyzePositions().catch(error => {
  console.error('❌ Error running analysis:', error);
  process.exit(1);
});

File: scripts/analyze_positions.py
#!/usr/bin/env python3
"""
Script to analyze open positions, compare DB vs Exchange, and find losing positions
"""

import json
import sys
import subprocess
from datetime import datetime
from collections import defaultdict

def run_sql_query(query):
    """Run SQL query and return results"""
    try:
        # Read database config from .env or use defaults
        result = subprocess.run(
            ['node', '-e', f'''
                const mysql = require("mysql2/promise");
                (async () => {{
                    const pool = mysql.createPool({{
                        host: process.env.DB_HOST || "localhost",
                        user: process.env.DB_USER || "root",
                        password: process.env.DB_PASSWORD || "",
                        database: process.env.DB_NAME || "bot_oc",
                        waitForConnections: true,
                        connectionLimit: 10
                    }});
                    try {{
                        const [rows] = await pool.execute(`{query}`);
                        console.log(JSON.stringify(rows));
                    }} finally {{
                        await pool.end();
                    }}
                }})();
            '''],
            capture_output=True,
            text=True,
            timeout=30
        )
        if result.returncode == 0:
            return json.loads(result.stdout)
        else:
            print(f"SQL Error: {result.stderr}", file=sys.stderr)
            return []
    except Exception as e:
        print(f"Error running SQL: {e}", file=sys.stderr)
        return []

def analyze_positions():
    """Main analysis function"""
    print("=" * 80)
    print("📊 POSITION ANALYSIS REPORT")
    print("=" * 80)
    print(f"Generated at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
    
    # 1. Get open positions from database
    print("1️⃣  Fetching open positions from database...")
    db_positions_query = """
        SELECT 
            p.id,
            p.bot_id,
            p.strategy_id,
            p.symbol,
            p.side,
            p.entry_price,
            p.amount,
            p.quantity,
            p.take_profit_price,
            p.stop_loss_price,
            p.pnl,
            p.pnl_percent,
            p.status,
            p.opened_at,
            p.created_at,
            b.exchange,
            b.binance_testnet,
            s.is_reverse_strategy,
            s.oc_threshold,
            s.extend,
            s.take_profit,
            s.stoploss
        FROM positions p
        LEFT JOIN bots b ON p.bot_id = b.id
        LEFT JOIN strategies s ON p.strategy_id = s.id
        WHERE p.status = 'open'
        ORDER BY p.opened_at DESC
    """
    
    db_positions = run_sql_query(db_positions_query)
    print(f"   Found {len(db_positions)} open positions in database\n")
    
    # 2. Analyze by exchange
    print("2️⃣  Analyzing by exchange...")
    by_exchange = defaultdict(list)
    for pos in db_positions:
        exchange = pos.get('exchange', 'unknown')
        is_testnet = pos.get('binance_testnet', False)
        key = f"{exchange}{'_testnet' if is_testnet else ''}"
        by_exchange[key].append(pos)
    
    for exchange, positions in by_exchange.items():
        print(f"   {exchange}: {len(positions)} positions")
    print()
    
    # 3. Analyze losing positions
    print("3️⃣  Analyzing losing positions...")
    losing_positions = [p for p in db_positions if p.get('pnl', 0) < 0]
    print(f"   Total losing positions: {len(losing_positions)}")
    
    if losing_positions:
        total_loss = sum(p.get('pnl', 0) for p in losing_positions)
        avg_loss = total_loss / len(losing_positions) if losing_positions else 0
        max_loss = min(p.get('pnl', 0) for p in losing_positions)
        
        print(f"   Total loss: {total_loss:.2f} USDT")
        print(f"   Average loss: {avg_loss:.2f} USDT")
        print(f"   Max loss: {max_loss:.2f} USDT")
        
        # Analyze by reason
        print("\n   Top losing positions:")
        sorted_losing = sorted(losing_positions, key=lambda x: x.get('pnl', 0))[:10]
        for i, pos in enumerate(sorted_losing, 1):
            print(f"   {i}. {pos.get('symbol')} {pos.get('side')} | "
                  f"Entry: {pos.get('entry_price', 0):.8f} | "
                  f"PnL: {pos.get('pnl', 0):.2f} USDT ({pos.get('pnl_percent', 0):.2f}%) | "
                  f"SL: {pos.get('stop_loss_price', 0) or 'N/A'}")
    print()
    
    # 4. Analyze winning positions
    print("4️⃣  Analyzing winning positions...")
    winning_positions = [p for p in db_positions if p.get('pnl', 0) > 0]
    print(f"   Total winning positions: {len(winning_positions)}")
    
    if winning_positions:
        total_profit = sum(p.get('pnl', 0) for p in winning_positions)
        avg_profit = total_profit / len(winning_positions) if winning_positions else 0
        max_profit = max(p.get('pnl', 0) for p in winning_positions)
        
        print(f"   Total profit: {total_profit:.2f} USDT")
        print(f"   Average profit: {avg_profit:.2f} USDT")
        print(f"   Max profit: {max_profit:.2f} USDT")
    print()
    
    # 5. Analyze by strategy type
    print("5️⃣  Analyzing by strategy type...")
    by_strategy_type = defaultdict(lambda: {'total': 0, 'winning': 0, 'losing': 0, 'total_pnl': 0})
    
    for pos in db_positions:
        is_reverse = pos.get('is_reverse_strategy', 0)
        strategy_type = 'COUNTER_TREND' if is_reverse else 'FOLLOWING_TREND'
        by_strategy_type[strategy_type]['total'] += 1
        pnl = pos.get('pnl', 0)
        by_strategy_type[strategy_type]['total_pnl'] += pnl
        if pnl > 0:
            by_strategy_type[strategy_type]['winning'] += 1
        elif pnl < 0:
            by_strategy_type[strategy_type]['losing'] += 1
    
    for strategy_type, stats in by_strategy_type.items():
        win_rate = (stats['winning'] / stats['total'] * 100) if stats['total'] > 0 else 0
        print(f"   {strategy_type}:")
        print(f"      Total: {stats['total']} | Winning: {stats['winning']} | Losing: {stats['losing']}")
        print(f"      Win Rate: {win_rate:.2f}% | Total PnL: {stats['total_pnl']:.2f} USDT")
    print()
    
    # 6. Analyze positions without SL
    print("6️⃣  Analyzing positions without Stop Loss...")
    no_sl_positions = [p for p in db_positions if not p.get('stop_loss_price') or p.get('stop_loss_price') == 0]
    print(f"   Positions without SL: {len(no_sl_positions)}")
    if no_sl_positions:
        print("   ⚠️  WARNING: These positions are at risk!")
        for pos in no_sl_positions[:5]:
            print(f"      - {pos.get('symbol')} {pos.get('side')} | "
                  f"Entry: {pos.get('entry_price', 0):.8f} | "
                  f"PnL: {pos.get('pnl', 0):.2f} USDT")
    print()
    
    # 7. Analyze positions without TP
    print("7️⃣  Analyzing positions without Take Profit...")
    no_tp_positions = [p for p in db_positions if not p.get('take_profit_price') or p.get('take_profit_price') == 0]
    print(f"   Positions without TP: {len(no_tp_positions)}")
    if no_tp_positions:
        print("   ⚠️  WARNING: These positions may not exit at profit target!")
        for pos in no_tp_positions[:5]:
            print(f"      - {pos.get('symbol')} {pos.get('side')} | "
                  f"Entry: {pos.get('entry_price', 0):.8f} | "
                  f"PnL: {pos.get('pnl', 0):.2f} USDT")
    print()
    
    # 8. Analyze by time opened
    print("8️⃣  Analyzing by time opened...")
    now = datetime.now()
    recent_positions = []
    old_positions = []
    
    for pos in db_positions:
        opened_at_str = pos.get('opened_at')
        if opened_at_str:
            try:
                opened_at = datetime.fromisoformat(str(opened_at_str).replace('Z', '+00:00'))
                hours_open = (now - opened_at.replace(tzinfo=None)).total_seconds() / 3600
                if hours_open < 24:
                    recent_positions.append((pos, hours_open))
                else:
                    old_positions.append((pos, hours_open))
            except:
                pass
    
    print(f"   Positions opened < 24h: {len(recent_positions)}")
    print(f"   Positions opened >= 24h: {len(old_positions)}")
    
    if old_positions:
        print("   ⚠️  WARNING: Old positions that may need review:")
        sorted_old = sorted(old_positions, key=lambda x: x[1], reverse=True)[:5]
        for pos, hours in sorted_old:
            print(f"      - {pos.get('symbol')} {pos.get('side')} | "
                  f"Open for {hours:.1f}h | "
                  f"PnL: {pos.get('pnl', 0):.2f} USDT")
    print()
    
    # 9. Summary
    print("=" * 80)
    print("📊 SUMMARY")
    print("=" * 80)
    total_pnl = sum(p.get('pnl', 0) for p in db_positions)
    print(f"Total Open Positions: {len(db_positions)}")
    print(f"Winning: {len(winning_positions)} | Losing: {len(losing_positions)}")
    print(f"Total PnL: {total_pnl:.2f} USDT")
    print(f"Win Rate: {(len(winning_positions) / len(db_positions) * 100) if db_positions else 0:.2f}%")
    print(f"Positions without SL: {len(no_sl_positions)}")
    print(f"Positions without TP: {len(no_tp_positions)}")
    print()
    
    # 10. Recommendations
    print("=" * 80)
    print("💡 RECOMMENDATIONS")
    print("=" * 80)
    
    if len(no_sl_positions) > 0:
        print("⚠️  CRITICAL: Some positions don't have Stop Loss!")
        print("   → Check PositionMonitor logs for SL placement errors")
    
    if len(no_tp_positions) > 0:
        print("⚠️  WARNING: Some positions don't have Take Profit!")
        print("   → Check PositionMonitor logs for TP placement errors")
    
    if len(losing_positions) > len(winning_positions):
        print("⚠️  WARNING: More losing positions than winning!")
        print("   → Review entry conditions and trend filters")
        print("   → Check if SL is being hit too early")
    
    if len(old_positions) > 0:
        print("⚠️  INFO: Some positions are open for > 24h")
        print("   → Review if these should still be open")
        print("   → Check trailing TP logic")
    
    print()
    print("=" * 80)

if __name__ == '__main__':
    analyze_positions()


